[Site Logo]: Add logo replace flow in Inspector controls

ntsekouras opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Resolves: https://github.com/WordPress/gutenberg/issues/49273

This PR adds the actions for set, replace and reset the site logo in Inspector controls. For users with no permissions to upload, if a logo exists it will show the 'chip' only.



## Testing Instructions
1. Insert a Site Logo and try the different flows



## Screenshots or screencast <!-- if applicable -->

https://user-images.githubusercontent.com/16275880/233666637-83a35da4-1094-4f37-9360-ad5418c6ae6d.mov

ntsekouras committed:
[Site Logo]: Add logo replace flow in Inspector controls

ntsekouras commented on packages/block-library/src/site-logo/editor.scss:
```diff
@@ -93,3 +93,65 @@
+.block-library-site-logo__inspector-upload-container,
+.block-library-site-logo__inspector-media-replace-container {
+	button.components-button {
+		color: $gray-900;
```
Most of these styles were copied from the `style variations` buttons. Should we consider a new 'variant' of button if we plan on reusing them? 🤔

jasmussen replied:
Reusing is my only remaining concern about this PR, because exatcly yes, we should have as little custom CSS as possible in PRs like these, and ideally reuse componentry whenever possible.

What would it require in this place? Do the style variation buttons need to be canonized as part of the Button style? Or is it a toggle? And if we canonize those can we reuse them here? If either answer is _yes_, it would be great to do that, either here or as a followup. Thanks so much for working on this.

ntsekouras replied:
In my last commit I removed some obsolete css, so it's way less now. Aside of that, these outline buttons seem to be a combination of `secondary` variant Buttons and a few overrides. I think it would make sense to have a specific Button variant for such styles.

Noting though that these styles can be used probably in just a `div` or something and it would be good to have a common styling - not sure where it should live in code though.. An example is this PR too, where we use similar styles(not the hover ones) for users that have no permissions to update, so we render the info in a `div`.

I think we can handle this in a follow up, but 🤷 . Whatever you prefer.

jasmussen commented:
Holy guacamole, that's impressively fast work! Looks almost ready to go. I'm just about to head out ofr the day, but I expect we can greenlight this quickly next week. Thank you.

mcsf commented on packages/block-library/src/site-logo/edit.js:
```diff
@@ -530,9 +582,44 @@ export default function LogoEdit( {
+								render={ ( { open } ) => (
+									<div className="block-library-site-logo__inspector-upload-container">
+										<Button onClick={ open }>
+											{ isLoading && <Spinner /> }
+											{ ! isLoading && label }
```
```suggestion
											{ isLoading ? <Spinner /> : label }
```

mcsf commented on packages/block-library/src/site-logo/edit.js:
```diff
@@ -530,9 +582,44 @@ export default function LogoEdit( {
+					{ canUserEdit && !! logoUrl && (
+						<SiteLogoReplaceFlow
+							{ ...mediaReplaceFlowProps }
+							name={ <InspectorLogoPreview /> }
```
It's slightly surprising that this just works; I guess it's more of a "label element" than a name.

The destination component, `MediaReplaceFlow`, doesn't type its props in the code but it does document them in its README file. It turns out that [the name is supposed to be a string](https://github.com/WordPress/gutenberg/blob/5c1d6ba207fce7c06ad94d51a62c204ae34d82f7/packages/block-editor/src/components/media-replace-flow/README.md#name). What should we do here? Leave that component untouched and work around? Or amend `MediaReplaceFlow` to explicitly accept elements? I wonder if it's safer to add a separate prop (like `label`) or just make `name` polymorphic. I have no opinion here; maybe the latter option is enough.

ntsekouras replied:
Oh yes great catch! I forgot to add the comment here yesterday and cc @ciampo and @mirka about this. Right now is as you say a workaround, without adding new props. We could deprecate 'name' and add a new polymorphic prop to handle this with a better name(no pun intended).

ciampo replied:
It's working because internally `name` is just interpolated in the JSX:

https://github.com/WordPress/gutenberg/blob/c6029852e1bd5af06aa87d55a6789620f7c6a727/packages/block-editor/src/components/media-replace-flow/index.js#L147-L155

Under a technical point of view, it feels safe to change the docs and allow `name` to be any `ReactNode`. I guess it depends more on whether you'd prefer to "limit" consumers of the component to pass strings of text for design / semantics / any other reasons.

mcsf reviewed (commented):
Looks pretty good. :)

ntsekouras committed:
Update packages/block-library/src/site-logo/edit.js

Co-authored-by: Miguel Fonseca <miguelcsf@gmail.com>

jasmussen commented:
Really really nice work, incredibly close right out of the gate. Thank you! This is virtually ready to land. Here's a GIF showing all the inspector cases for the new media button:

![site logo](https://user-images.githubusercontent.com/1204802/233920210-1593cee4-428b-41c2-9833-d64c5e628af4.gif)

A few small things:

<img width="351" alt="Screenshot 2023-04-24 at 08 48 31" src="https://user-images.githubusercontent.com/1204802/233920241-bae8a413-0d72-4354-867b-db4d89290ce2.png">

This should either say "Add media" (I could see us allowing a video site logo in the future) or "Add image" if we want to postpone the more generic term.

<img width="371" alt="Screenshot 2023-04-24 at 08 48 50" src="https://user-images.githubusercontent.com/1204802/233920350-cb61f71d-e848-4940-a537-2d3de3780f1e.png">

The button should be 40px tall (indeed we should be deprecating 36px, see #46734). There should be a prop to leverage this.

In the mockup, I have image width grouped with "Media", but I'm waffling on that one. Since it's already in the Settings panel, maybe we keep that and revisit this aspect when we can refresh the Image inspector?

<img width="338" alt="Screenshot 2023-04-24 at 08 49 17" src="https://user-images.githubusercontent.com/1204802/233920775-9616260d-6025-4f2d-bfc5-503e167bf744.png">

In the mockups, I imagined also seeing the file extension of the file:

<img width="402" alt="Screenshot 2023-04-24 at 08 53 02" src="https://user-images.githubusercontent.com/1204802/233920848-763b1e6f-369d-4504-ba1b-d40916491d5b.png">

To me that's actually about indicating what that field is about, actual file management. It's not critical we get this right, especially if this is a challenge, but I think it would be a nice improvement.

The media field that holds the filename and circular thumbnail is meant to be an ItemGroup, because theoretically in the future it could become a layer management interface.

<img width="318" alt="Screenshot 2023-04-24 at 08 56 11" src="https://user-images.githubusercontent.com/1204802/233921389-ea32b55c-a126-4f96-8ef3-159cfd1b52a0.png">

So conceptually it's similar to other ItemGroups, like the recently landed Duotone:

<img width="334" alt="Screenshot 2023-04-24 at 08 55 06" src="https://user-images.githubusercontent.com/1204802/233921436-c987232c-0923-4e1a-a1d0-f98fe1917a07.png">

I don't know how many styles we get for free by actually using the ItemGroup here, but hopefully the metrics. The item should be 40px tall:

<img width="466" alt="Screenshot 2023-04-24 at 08 54 06" src="https://user-images.githubusercontent.com/1204802/233921500-ddbe7abb-acd4-45af-ad9f-50199949002b.png">

The thumbnail should be 20x20:

<img width="268" alt="Screenshot 2023-04-24 at 08 54 14" src="https://user-images.githubusercontent.com/1204802/233921714-9335deea-0846-40b1-b027-9825ae360ca4.png">

And it should have a small gray overlaid semitransparent border, like color swatches do:

<img width="329" alt="Screenshot 2023-04-24 at 08 58 15" src="https://user-images.githubusercontent.com/1204802/233921772-ca2c8825-0267-4229-bd65-d68136f5a1da.png">

Hopefully using the components should net all those for "free".

Nice work. Small fixes, and IMO this is ready to land.

jameskoster commented:
It might be good to update the featured image inspector UI to match this as a follow-up?

jasmussen commented:
Yes, good enhancement. And Cover, honestly. Cover might need a little design.

ntsekouras committed:
feedback

ntsekouras commented:
I think I addressed the feedback. Let me know if anything more is needed.

> In the mockup, I have image width grouped with "Media", but I'm waffling on that one. Since it's already in the Settings panel, maybe we keep that and revisit this aspect when we can refresh the Image inspector?

I agree. We can follow up on that one.

Mamaduka commented on packages/block-library/src/site-logo/edit.js:
```diff
@@ -461,17 +489,48 @@ export default function LogoEdit( {
+		} );
+	};
+
+	const InspectorLogoPreview = ( { itemGroupProps } ) => {
```
Can we define this component outside of the Edit component? So we don't have to re-create it on every rerender. I believe that's the recommended practice in React.

ntsekouras committed:
define `InspectorLogoPreview` outside

jasmussen commented:
Thank you for such fast work! This is really close, and could honestly land as is. Small note, `Add media`, not `Add Media`.

There are a couple of nitpicky questions that mostly relate to the componentry and how it's used, but it reveals itself primarily in how these hover styles present themselves. For example in the empty state, the hover state of media button is animated and makes the outline thicker, making it appear as if it's focused:

![hover](https://user-images.githubusercontent.com/1204802/233996003-e5583f43-06a9-4b28-a60d-4b790c622cd2.gif)

For that style of button (gray-bordered), that's not the correct style compared to the style variation buttons which also have the gray border. 

Related to that, the filled in media button also has this hover style:

![hover media item](https://user-images.githubusercontent.com/1204802/233996126-9cc412a3-86fe-4f2c-9afa-9b48400ebe40.gif)

But ItemGroup items do not:

![itemgroup hover](https://user-images.githubusercontent.com/1204802/233996147-32a7d569-31a4-4549-b3f6-5a0520294938.gif)

Overall, we should unify on a single hover style that's the same across outline buttons an itemgroups aside, which is mainly about a blue hover style. It can also be a blue outline, like the Style variation buttons, but they should not be thicker like the focus style. 

So my main question is: where are these hover styles inherited from? Ideally we shouldn't need any custom CSS at all for these components. I'm not personally averse to it, but just noting that our components should be able to handle these things natively, so that if one day we advance the hover style of the buttons, they are advanced across all instances.

ntsekouras committed:
change label case

ntsekouras committed:
adjust styles

ntsekouras commented:
`ItemGroup` is a component to group items and the examples shared above have different inner components. That's why the styles don't match. There seems to be the styles we override for `styles button` also use the `secondary` variant and it would include more changes in components package. I think we should handle this separately in a different PR. 

For now I updated the styles introduced here to match the `styles` button.

jasmussen reviewed (approved):
Visually and on its behavior, this one is ready:

![status](https://user-images.githubusercontent.com/1204802/234218223-55d9e7bf-7c68-4bb8-90e8-1c694e0bfaa8.gif)

I shared some general notes on componentry reuse here: https://github.com/WordPress/gutenberg/pull/49992#discussion_r1176148509 — I will defer to devs on the componentry on how when and where to unify here. It's a fuzzy rule: generally we should not write custom CSS for pieces like these, as it's all meant to be reused base componentry. But I wouldn't block this PR for it. We just need to all pull in that direction so that if we were to, say, update the border style on the componentry level, all linked components would inherit the new styles, instead of having local overrides that causes drift.

ntsekouras committed:
update `MediaReplaceFlow` readme and use valid html inside `button`

Mamaduka reviewed (approved):
Excellent work, Nik!

ntsekouras committed:
update hover/focus styles

mtias commented:
This is a great improvement, thanks for working on it!

paaljoachim commented:
I added a comment here in the closed issue: https://github.com/WordPress/gutenberg/issues/49273#issuecomment-1528967005
In regards to expanding this media control to also other blocks such as Cover and Group.

ndiego commented:
@ntsekouras @jasmussen I was just testing this for 15.7 and found that if the file name for the image is really long, the image will get constrained below 20px. We may want to fix ahead of the 15.7 release on Wednesday. 

cc @bph 

<img width="346" alt="image" src="https://user-images.githubusercontent.com/4832319/235353048-2131bb8c-c420-4768-9504-dc3caa9a3bef.png">
